File: cpp/image-convolution/convolution.cpp
#include "convolution.h"

#include <thread>

using namespace std;

color_t roundToColor(const float& value) {
	return value + .5;
}

const color_t& clamp(const color_t& value, const color_t& min, const color_t& max) {
	if (value < min)
		return min;
	else if (value > max)
		return max;
	else
		return value;
}

void convolveWorker(
	const Image& input, Image& output, const Kernel<>& kernel, int xOffset, int xSkip
) {
	const static Pixel<> FALLBACK{};
	int radius = kernel.radius;
	color_t maximumColorValue = input.maximumColorValue;
	Pixel<float> unroundedPixel;
	// Iterate over image
	for (int x = xOffset; x < input.width; x += xSkip) {
		for (int y = 0; y < input.height; ++y) {
			unroundedPixel.fill(0);

			// Iterate over kernel
			for (int ox = -radius; ox <= radius; ++ox) {
				for (int oy = -radius; oy <= radius; ++oy) {
					const float kernelValue = kernel.at(ox + radius, oy + radius);
					const Pixel<>& imageValue = input.atOr(x + ox, y + oy, FALLBACK);
					// Iterate over channels
					for (int channel = 0; channel < 3; ++channel) {
						unroundedPixel[channel] += kernelValue * imageValue[channel];
					}
				}
			}

			Pixel<>& outputPixel = output.at(x, y);
			for (int channel = 0; channel < 3; ++channel) {
				const color_t rounded = roundToColor(unroundedPixel[channel]);
				outputPixel[channel] = clamp(
					rounded, static_cast<color_t>(0), maximumColorValue
				);
			}
		}
	}
}

void convolve(const Image& input, Image& output, const Kernel<>& kernel) {
	unsigned int workerCount = std::thread::hardware_concurrency();

	vector<thread> threads;
	threads.reserve(workerCount);
	for (unsigned int i = 0; i < workerCount; i++)
		threads.emplace_back(
			convolveWorker,
			cref(input),
			ref(output),
			cref(kernel),
			i,
			workerCount
		);
	for (thread& thread : threads)
		thread.join();
}

File: rust/image-convolution/src/convolution.rs
pub mod convolution {
	use std::sync::Arc;
	use std::thread;

	use crate::image::image::Image;
	use crate::kernel::kernel::Kernel;

	pub fn convolve(input: Arc<Image>, output: Arc<Image>, kernel: Arc<Kernel>) {
		let worker_count = num_cpus::get();
		let mut handles = Vec::new();
		handles.reserve_exact(worker_count);
		for i in 0..worker_count {
			let input = input.clone();
			let output = output.clone();
			let kernel = kernel.clone();
			let handle = thread::spawn(move || {
				convolve_worker(input, output, kernel, i, worker_count);
			});
			handles.push(handle);
		}
		for handle in handles {
			handle.join().unwrap();
		}
	}

	fn convolve_worker(
		input: Arc<Image>,
		output: Arc<Image>,
		kernel: Arc<Kernel>,
		x_offset: usize,
		x_skip: usize,
	) {
		let fallback = [0, 0, 0];
		let radius = kernel.radius as isize;
		let mut unrounded_pixel: [f32; 3];
		let maximum_color_value = input.maximum_color_value as f32;
		for x in (x_offset..input.width).step_by(x_skip) {
			for y in 0usize..input.height {
				unrounded_pixel = [0f32, 0f32, 0f32];

				for ox in -radius..=radius {
					for oy in -radius..=radius {
						let kernel_value = kernel.get((ox + radius) as usize, (oy + radius) as usize);
						let image_value = input
							.get(x as isize + ox, y as isize + oy)
							.unwrap_or(&fallback);
						for channel in 0..3 {
							unrounded_pixel[channel] += kernel_value * image_value[channel] as f32;
						}
					}
				}

				let pixel = output.index_mut_unsafe(x, y);
				for channel in 0..3 {
					let unrounded = unrounded_pixel[channel];
					let rounded = unrounded.round();
					let clamped = clamp(rounded, 0f32, maximum_color_value);
					let casted = clamped as u8;
					pixel[channel] = casted;
				}
			}
		}
	}

	fn clamp<T>(value: T, min: T, max: T) -> T where T: PartialOrd {
		if value > max {
			return max;
		} else if value < min {
			return min;
		} else {
			return value;
		}
	}
}